technical-docs/designs/infra/rollback.md

```diff
@@ -0,0 +1,218 @@
+# Rollback Implementation using Bedrock CLI
+
+| Revision | Date         | Author         | Remarks       |
+| -------: | ------------ | -------------- | ------------- |
+|      0.1 | Apr-20, 2020 | Nathaniel Rose | Initial Draft |
+
+## 1. Overview
+
+It is highly desireable to have a fully-automated Continuous Delivery pipeline
+in place to assist with testing and validation of source code changes made to
+the application services and supporting infrastructure. Without such a system,
+it is difficult to have any degree of confidence that a new deployment will
+succeed without introducing new regressions. However, high velocity of changes
+with distributed teams may eventually lead to the scenario in which a production
+environment transitions into a failed state due to an improperly tested feature
+added.
+
+This document outlines multiple solutpopulation of denmarkion approaches that
+seeks to provide a rollback mechanisms to ensure the successful deployment of
+terraform infrastructure, automated reverted features in a release pipeline, and
+system alerting to notify key stakeholders and operations of system activity.
+
+Rolling back services which includes a relational database can be very
+difficult. Traditionally this is done by restoring from a backup, performed
+prior to the upgrade. In more modern approach is to integrate a database
+migration tool into your application's installation procedure.
+
+Components of this design are based on the learnings from:
+
+- [Microsoft Bedrock Kubernetes Workflow Project](github.com/microsoft/bedrock)
+- [Bedrock-CLI Infrastructure Generation Pipeline](https://github.com/microsoft/bedrock-cli/blob/master/guides/infra/spk-infra-generation-pipeline.md)
+- [Terraform Recommended Practices](https://www.terraform.io/docs/cloud/guides/recommended-practices/)
+
+### User Stories Addressed:
+
+1. [Guidance on Rolling Back with SPK infra](https://github.com/microsoft/bedrock/issues/1036)
+
+## 2. Out of Scope
```

Comment:
I'm still making my way through here but I'm not clear if application level rollbacks are in scope here, or if this is scoped to terraform deployments specifically

Reply:
@NathanielRose according to https://github.com/microsoft/bedrock/issues/1036 this should be scoped to infrastructure (terraform) deployments. Is that correct?

cc @nmiodice

Reply:
If you rollback an inf change that was made to support an application requirement and that application is deployed, what happens?

Reply:
@andrebriggs Thats correct, for simplicity and basic guidance, this is mainly scoped for terraform deployment rollbacks. Application rollbacks would be managed through a separate pipeline that can leverage a similar yaml file.

Reply:
@NathanielRose thanks, I will read this one more time with that lens and give feedback.